src/gui/render/graphics_shaders/shader_program.cpp

```diff
@@ -0,0 +1,278 @@
+
+        mesh->release();
+    }
+}
+
+void
+ShaderProgram_Elements::render(
+    screen_size_t width, screen_size_t height, GLuint framebuffer_ID
+) {
+#if DEBUG()
+    static bool has_logged = false;
+    if (data.size() == 0 && !has_logged) {
+        LOG_WARNING(
+            logging::opengl_logger, "Nothing to be rendered. Program id {}",
+            opengl_program_.get_program_ID()
+        );
+        has_logged = true;
+    }
+#endif
+
+    Render_Base::render(width, height, framebuffer_ID);
+
+    for (const auto mesh : data) {
+        if (!mesh->do_render()) {
+            continue;
+        }
+
+        mesh->bind();
+
+        // test if T inherits from Instancing or not
+
+        auto num_vertices = mesh->get_num_vertices();
+        auto element_type = mesh->get_element_type();
+
+        // Draw the triangles !
+        glDrawElements(
+            GL_TRIANGLES,                      // mode
+            num_vertices,                      // count
+            static_cast<GLenum>(element_type), // type
+            (void*)0                           // element array buffer offset
+        );
+
+        mesh->release();
+    }
+}
+
+void
+ShaderProgram_Instanced::render(
+    screen_size_t width, screen_size_t height, GLuint framebuffer_ID
+) {
+#if DEBUG()
+    static bool has_logged = false;
+    if (data.size() == 0 && !has_logged) {
+        LOG_WARNING(
+            logging::opengl_logger, "Nothing to be rendered. Program id {}",
+            opengl_program_.get_program_ID()
+        );
+        has_logged = true;
+    }
+#endif
+
+    Render_Base::render(width, height, framebuffer_ID);
+
+    for (const auto mesh : data) {
+        if (!mesh->do_render()) {
+            continue;
+        }
+
+        mesh->bind();
+
+        // Draw the triangles !
+        glDrawArraysInstanced(
+            GL_TRIANGLE_STRIP,        // mode
+            0,                        // start
+            mesh->get_num_vertices(), // number of vertices
+            mesh->get_num_models()    // number of models
+
+        );
+
+        mesh->release();
+    }
+}
+
+void
+ShaderProgram_ElementsInstanced::render(
+    screen_size_t width, screen_size_t height, GLuint framebuffer_ID
+) {
+#if DEBUG()
+    static bool has_logged = false;
+    if (data.size() == 0 && !has_logged) {
+        LOG_WARNING(
+            logging::opengl_logger, "Nothing to be rendered. Program id {}",
+            opengl_program_.get_program_ID()
+        );
+        has_logged = true;
+    }
+#endif
+
+    Render_Base::render(width, height, framebuffer_ID);
+
+    for (const auto mesh : data) {
+        if (!mesh->do_render()) {
+            continue;
+        }
+
+        //            assert(static_cast<GLsizei>(mesh->get_num_vertices()) >= 0 &&
+        //            "Num vertices must non-negative");
+        //            assert(static_cast<GLsizei>(mesh->get_num_models()) >= 0 &&
+        //            "Num models must non-negative");
+
+        mesh->bind();
```

Comment:
You should make an RAII wrapper to do bind/release automatically. See scoped_lock for an example

Reply:
No, this is already twice abstracted, and unbind doesn't really do anything.

Reply:
If `unbind()` doesn't do anything, why is it called?

Reply:
Eh ok fine.